File: app/src/psd.ts
import { ColorModeDataSection } from './sections/colorModeDataSection';
import { ImageDataSection } from './sections/imageDataSection';
import { ImageResourcesSection } from './sections/imageResourcesSection';
import { LayerAndMaskInformationSection } from './sections/layerAndMaskInformation';

export class Psd {
  // fileHeaderSection: FileHeaderSection;
  colorModeDataSection: ColorModeDataSection;
  imageResourcesSection: ImageResourcesSection;
  layerAndMaskInformationSection: LayerAndMaskInformationSection;
  ImageDataSection: ImageDataSection;

  constructor(view: DataView) {
    // File Header Section
    // https://www.adobe.com/devnet-apps/photoshop/fileformatashtml/#50577409_pgfId-1055726
    // this.fileHeaderSection = new FileHeaderSection(binary);

    // Color Mode Data Section
    const colorModeDataSectionStart = 26;
    this.colorModeDataSection = new ColorModeDataSection(
      view,
      colorModeDataSectionStart
    );

    // Image Resources Section
    const imageResourcesSectionStart =
      colorModeDataSectionStart + this.colorModeDataSection.length;
    this.imageResourcesSection = new ImageResourcesSection(
      view,
      imageResourcesSectionStart
    );

    // Layer and Mask Information Section
    const layerAndMaskInformationSectionStart =
      imageResourcesSectionStart + this.imageResourcesSection.length;
    this.layerAndMaskInformationSection = new LayerAndMaskInformationSection(
      view,
      layerAndMaskInformationSectionStart
    );

    // Image Data Section
    const imageDataSectionStart =
      layerAndMaskInformationSectionStart +
      this.layerAndMaskInformationSection.length;
    this.ImageDataSection = new ImageDataSection(view, imageDataSectionStart);
  }
}

File: app/src/psd.rs
mod color_mode_data_section;
mod file_header_section;
mod image_resources_section;
mod layer_and_mask_information_section;

use self::{
    color_mode_data_section::{get_color_mode_data_section, ColorModeDataSection},
    file_header_section::{get_file_header_section, FileHeaderSection},
    image_resources_section::{get_image_resources_section, ImageResourcesSection},
    layer_and_mask_information_section::{
        get_layer_and_mask_information_section, LayerAndMaskInformationSection,
    },
};
use crate::binary::Binary;

pub struct Psd {
    pub file_header_section: FileHeaderSection,
    pub color_mode_data_section: ColorModeDataSection,
    pub image_resources_section: ImageResourcesSection,
    pub layer_and_mask_information_section: LayerAndMaskInformationSection,
}

pub fn get_psd(binary: &mut Binary) -> Psd {
    Psd {
        file_header_section: get_file_header_section(binary),
        color_mode_data_section: get_color_mode_data_section(binary),
        image_resources_section: get_image_resources_section(binary),
        layer_and_mask_information_section: get_layer_and_mask_information_section(binary),
    }
}
